pkg/gardenlet/controller/vpaevictionrequirements/reconciler.go

```diff
@@ -0,0 +1,164 @@
+	isNowInMaintenanceTimeWindow := maintenanceTimeWindow.Contains(r.Clock.Now())
+
+	if isNowInMaintenanceTimeWindow {
+		log.Info("Shoot is inside maintenance window, removing the EvictionRequirement to allow downscaling", "shoot-namespace", vpa.GetNamespace(), "maintenanceWindow", maintenanceTimeWindow)
+		_, err := controllerutils.GetAndCreateOrMergePatch(seedCtx, r.SeedClient, vpa, func() error {
+			if !hasUpscaleOnlyEvictionRequirement(vpa.Spec.UpdatePolicy.EvictionRequirements) {
+				return nil
+			}
+			vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{}
+			return nil
+		})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// requeue when the maintenance window ends, such that we can add the EvictionRequirement again
+		endTime := maintenanceTimeWindow.AdjustedEnd(r.Clock.Now())
+		requeueAfter := endTime.Sub(r.Clock.Now())
+		log.Info("Requeueing VPA", "requeueAfter", requeueAfter)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
+	} else {
+		log.Info("Shoot is not inside maintenance window, adding EvictionRequirement to deny downscaling", "shoot-namespace", vpa.GetNamespace(), "maintenanceWindow", maintenanceTimeWindow)
+		_, err := controllerutils.GetAndCreateOrMergePatch(seedCtx, r.SeedClient, vpa, func() error {
+			if hasUpscaleOnlyEvictionRequirement(vpa.Spec.UpdatePolicy.EvictionRequirements) {
+				return nil
+			}
+			vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{
+				{
+					Resources:         cpuAndMemory,
+					ChangeRequirement: vpaautoscalingv1.TargetHigherThanRequests,
+				},
+			}
+			return nil
+		})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// requeue when the next maintenance window begins, such that we can remove the EvictionRequirement
+		nextWindowBegin := maintenanceTimeWindow.AdjustedBegin(r.Clock.Now())
+		if nextWindowBegin.Before(r.Clock.Now()) {
+			nextWindowBegin = nextWindowBegin.AddDate(0, 0, 1)
+		}
+		requeueAfter := nextWindowBegin.Sub(r.Clock.Now())
+		log.Info("Requeueing VPA", "requeueAfter", requeueAfter)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
+	}
+}
+
+func (r *Reconciler) reconcileVPAForDownscaleDisabled(seedCtx context.Context, vpa *vpaautoscalingv1.VerticalPodAutoscaler) (reconcile.Result, error) {
+	log := logf.FromContext(seedCtx)
+	log.Info("Adding EvictionRequirement to deny downscaling")
+
+	_, err := controllerutils.GetAndCreateOrMergePatch(seedCtx, r.SeedClient, vpa, func() error {
+		if !hasUpscaleOnlyEvictionRequirement(vpa.Spec.UpdatePolicy.EvictionRequirements) {
+			vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{
+				{
+					Resources:         []corev1.ResourceName{corev1.ResourceCPU, corev1.ResourceMemory},
+					ChangeRequirement: vpaautoscalingv1.TargetHigherThanRequests,
+				},
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	log.Info("Not requeueing VPA")
```

Comment:
Similar to above, the log messages have no identifying component. Please either pass in namespace/name or create the logger beforehand to always include those.

Reply:
done.